### PR TITLE
Consume user activation in `element.requestFullscreen()`

### PR DESCRIPTION
**Pull request by foolip**
Depends on https://github.com/whatwg/html/pull/3851.

Fixes https://github.com/whatwg/fullscreen/issues/152.

- [x] At least two implementers are interested (and none opposed):
   * Chromium: https://github.com/whatwg/fullscreen/pull/153#issuecomment-496521939
   * Gecko: https://github.com/whatwg/fullscreen/pull/153#issuecomment-1298746998
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/36969
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1804635
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=247920
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/22303

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/fullscreen/153.html" title="Last updated on May 3, 2023, 7:23 AM UTC (e1ab822)">Preview</a> | <a href="https://whatpr.org/fullscreen/153/a69e295...e1ab822.html" title="Last updated on May 3, 2023, 7:23 AM UTC (e1ab822)">Diff</a>

### CODE REVIEW

**Comment by mustaqahmed** [comment]
Thanks, this looks good from user activation perspective.

**Comment by SummerLW** [comment]
https://github.com/web-platform-tests/wpt/pull/16758 has been merged into wpt repository.
Three new tests about this behavior are now in https://w3c-test.org/fullscreen/api/
element-request-fullscreen-twice-manual.tentative.html
element-request-fullscreen-two-elements-manual.tentative.html
element-request-fullscreen-two-iframes-manual.tentative.html

File: fullscreen.bs
Comment (mustaqahmed) [inline fullscreen.bs]: FYI: This should now point to the newly added definition of [consumption](https://html.spec.whatwg.org/multipage/interaction.html#consume-user-activation).  Not sure why the PR preview doesn't show a link.

Comment (mustaqahmed) [reply fullscreen.bs:273]: Looks like the preview is correctly pointing to the HTML spec link now.

**Review by mustaqahmed — COMMENTED** [review]
I wanted to approve this PR, but apparently I don't have permission to do that!  @domenic?

File: fullscreen.bs
Comment (domenic) [inline fullscreen.bs:274]: This should be pendingDoc's relevant global object, not its browsing context.

**Review by domenic — CHANGES_REQUESTED** [review]

**Comment by domenic** [comment]
@foolip, can we help move this along? I'm happy to rebase it and also take care of any changes after #208.

This would probably solve the issue @marcoscaceres raises in https://github.com/web-platform-tests/wpt/issues/36727 of giving a cross-browser way of consuming user activation.

**Comment by marcoscaceres** [comment]
I guess it depends on what features/API calls follow a call to `.requestFullScreen()`... I see there is mention of Pointerlock in the bug and Screen Orientation Lock is another.  

I'm simultaneously hoping we can prevent the Screen Orientation API (which depends on FS API) from arbitrarily changing the screen orientation [without a user gesture](https://github.com/w3c/screen-orientation/pull/218). 

If you have an Android device handy, try this out in Chrome: 
https://marcoscaceres.github.io/playground/annoying.html 

(if you can't try it out, it loops rapidly changing the screen orientation making it difficult of users to access the browsers UI - you need to either turn off the screen or quickly swipe from the bottom to stop it). 

I personally don't think the above behavior should be acceptable for any API to do on the Web. So, I was hoping we wouldn't consume the user activation in FS API, but rather have the Screen Orientation API consume the user activation. 

Alternatively, we could have it consume the user activation if #186 was in place, all in one go. The advantage being that it wouldn't breaking any sites that are depending on FS API to not consuming the user activation.

For pointerlock, maybe a solution like #186 could also work. You request them all at once, and exiting fulls screen releases all the locks.

**Comment by EdgarChen** [comment]
Currently, Gecko also consume the user activation as we had received a bug of trapping the user in fullscreen mode, which invites DOS and spoofing possibilities. And for pointer lock, we allow it without a vaild user activation if document is in fullscreen mode.

**Comment by marcoscaceres** [comment]
Discussing with my WebKit colleagues... but if both Gecko and Chromium consume it, we may follow suit. Let me get back to you. 

> And for pointer lock, we allow it without a vaild user activation if document is in fullscreen mode.

That is contrary to what is in the Pointer Lock spec, right? Why don't we just link the actions together via a mechanism like in #186? 

(also, we should really rewrite the pointer lock spec to integrate properly with the Full Screen API)

**Comment by mustaqahmed** [comment]
> Currently, Gecko also consume the user activation as we had received a bug of trapping the user in fullscreen mode, which invites DOS and spoofing possibilities.

Chrome originally proposed this consumption behavior to prevent a spoof, see https://crbug.com/851302.

>> And for pointer lock, we allow it without a vaild user activation if document is in fullscreen mode.
>
> That is contrary to what is in the Pointer Lock spec, right? Why don't we just link the actions together via a mechanism like in https://github.com/whatwg/fullscreen/issues/186?

IIRC Chrome added a similar hack to support some compat argument, can't recall any details!  Yes, #186 would be nice.

**Comment by EdgarChen** [comment]
https://github.com/whatwg/fullscreen/issues/186 sounds reasonable to me.

**Comment by marcoscaceres** [comment]
@EdgarChen, if we go with something like #207 (new dictionary member), would you be willing to remove the pointerLock workaround in Firefox? That way, we can just have a consistent model across three specs. 

Similar question for Chrome folks. I'm willing to spec that up if everyone is supportive.

File: fullscreen.bs
Comment (domenic) [reply fullscreen.bs:270]: Use `<a for=navigable>activate window</a>`

**Comment by EdgarChen** [comment]
> @EdgarChen, if we go with something like #207 (new dictionary member), would you be willing to remove the pointerLock workaround in Firefox?

To clarify this a bit more: This is not a workaround added for consuming user activation, we behave the same all the way back to Firefox 22 at least (before the current user activation model in spec is proposed). I am worried that removing this would cause compatibility problem. :(

**Comment by foolip** [comment]
@marcoscaceres there are 3 tests in wpt that are tentative due to #152:

https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/element-request-fullscreen-twice-manual.tentative.html
https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/element-request-fullscreen-two-elements-manual.tentative.html
https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/element-request-fullscreen-two-iframes-manual.tentative.html

But I think with consuming, this can be reduced to just the first test, ensuring the second call is rejected.

File: fullscreen.bs
Comment (domenic) [reply fullscreen.bs:270]: Yeah, looks like that's not properly exported. I'll get on it.

**Comment by marcoscaceres** [comment]
Filed WebKit bug:
https://bugs.webkit.org/show_bug.cgi?id=247920

Thanks for the details about the tests, @foolip. If those can get updated and changed to non-tentative, that would be great.

**Comment by upsuper** [comment]
I have a feeling that there is nothing wrong to allow `requestFullscreen` without user activation when the document is already in fullscreen, or is entering fullscreen (by a previous call). Though to be honest I'm not sure whether it is useful to allow that either.

**Comment by foolip** [comment]
@upsuper I think you're right about that, since the page script already has full control over what's being presented and could replace all the contents of the current fullscreen element. However, that'd be a change from the currently implemented behavior, and I'd prefer to handle that separately from this. Can you file an issue if you think we should change that?

**Comment by foolip** [comment]
@marcoscaceres test PR is up and the whole PR template is filled in now.

**Comment by marcoscaceres** [comment]
Thanks @foolip! Everything looks great.

File: fullscreen.bs
Comment (domenic) [reply fullscreen.bs:270]: Actually you should just use `<var>pendingDoc</var>'s <a>relevant global object</a>`! No need to go all the way to navigable and back.

Comment (marcoscaceres) [inline fullscreen.bs:270]:
```suggestion
 <li><p>If <var>error</var> is false, then <a>consume user activation</a> given
 <var>pendingDoc</var>'s <a>relevant global object</a>.
```

**Comment by marcoscaceres** [comment]
This is now also implemented in WebKit. 

@foolip, anything left to get this merged?

File: fullscreen.bs
Comment (foolip) [reply fullscreen.bs]: Ah, thanks!

Comment (foolip) [reply fullscreen.bs]: Applied, thanks!

**Comment by foolip** [comment]
@marcoscaceres sorry I totally dropped the ball here. I've updated the PR and all I need now is review from someone with write access. Paging @annevk, as @domenic seems to be OOO today.

**Review by annevk — APPROVED** [review]
This seems fine. The step immediately preceding this could do with some love at some point. The booleans are still hurting me.

**Comment by foolip** [comment]
Thanks @annevk! Let me try inverting the preceding steps and see if you like that.

**Pull request merged**
